pkg/driver/node/taint.go

```diff
@@ -0,0 +1,220 @@
+/*
+Copyright 2022 The Kubernetes Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package node
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8stypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+)
+
+const (
+	AgentNotReadyNodeTaintKey = "s3.csi.aws.com/agent-not-ready" // key of taints to be removed on driver startup
+
+	// CSI driver name for registration verification
+	csiDriverName = "s3.csi.aws.com"
+
+	// TaintWatcherDuration is the maximum duration for the not-ready taint watcher to run.
+	TaintWatcherDuration = 1 * time.Minute
+)
+
+// Struct for JSON patch operations
+type JSONPatch struct {
+	OP    string      `json:"op,omitempty"`
+	Path  string      `json:"path,omitempty"`
+	Value interface{} `json:"value"`
+}
+
+// StartNotReadyTaintWatcher launches a short-lived Node informer that removes the
+// s3.csi.aws.com/agent-not-ready taint. The informer is stopped after maxWatchDuration.
+func StartNotReadyTaintWatcher(clientset kubernetes.Interface, nodeID string, maxWatchDuration time.Duration) {
+	if nodeID == "" {
+		klog.V(4).Infof("nodeID is empty, skipping taint watcher")
+		return
+	}
+
+	klog.Infof("Starting taint watcher for node %s (max duration: %v)", nodeID, maxWatchDuration)
```

Comment:
Might be useful to also print the taint name we're looking for (i.e., `AgentNotReadyNodeTaintKey`)